Fixing the GetInterfaceMap tests for mono in the runtime repo.

bholmes opened the pull request:
Addressing 4 issues for GetInterfaceMap and default interface methods
 - Only methods marked as virtual on on interface should be added to the
   interface map. (no static or instance)
 - If the found target method is ambiguous (a diamond) the target is
   null.
 - If the found target method's class in an interface, then the target
   class is the interface class, else it is the class of the RuntimeType
   (aka this)
 - If the found target method is abstract (reabstraction) then the
   target is null.

https://github.com/dotnet/runtime/blob/6072e4d/src/tests/reflection/DefaultInterfaceMethods/GetInterfaceMapConsumer.cs

joshpeterson commented:
It turns out that I did not need to make any changes to Mono in the `support-netstandard-2-1` branch that are actually required for .NET Standard 2.1 support in Mono. The changes there are only for .NET FW 4.8 support via the profile stubber. So I think this pull request should instead target `unity-master-new-unitychanges`.

joncham commented:
To be clear, are these changes all our own? Or is the taken from dotnet repo version of Mono?

bholmes commented:
> It turns out that I did not need to make any changes to Mono in the `support-netstandard-2-1` branch that are actually required for .NET Standard 2.1 support in Mono. The changes there are only for .NET FW 4.8 support via the profile stubber. So I think this pull request should instead target `unity-master-new-unitychanges`.

OK I will retarget if and when I merge this. Retargeting here will pull in your changes as well.

joshpeterson commented:
> Retargeting here will pull in your changes as well.

Oh, good point. Ok then, sounds good.

bholmes commented:
> To be clear, are these changes all our own? Or is the taken from dotnet repo version of Mono?

They are ours. I am attempting a discussion to see if and how this goes upstream. I think it should go to dotnet/mono and mono/mono as FWICT this code has not changed.

bholmes commented:
I want to check the order on this. I think we need to check for conflicts first before abstract.  But found methods on a class can be checked first as that should not be in conflict ever and we can save that time?

bholmes committed:
Fixing the GetInterfaceMap tests for mono in the runtime repo.

Addressing 4 issues for GetInterfaceMap and default interface methods

Only methods marked as virtual on on interface should be added to the
interface map. (no static or instance)
If the found target method is ambiguous (a diamond) the target is
null.
If the found target method's class in an interface, then the target
class is the interface class, else it is the class of the RuntimeType
(aka this)
If the found target method is abstract (reabstraction) then the
target is null.

https://github.com/dotnet/runtime/blob/6072e4d/src/tests/reflection/DefaultInterfaceMethods/GetInterfaceMapConsumer.cs
